src/main/java/org/candlepin/model/ContentCurator.java

```diff
@@ -645,63 +645,50 @@ public Set<Content> getChildrenContentOfProductsByUuids(Collection<String> produ
+                    FROM cp_product_provided_products
+                ),
+                product_graph(uuid, depth) AS (
+                    SELECT CAST(pool.product_uuid AS CHAR(32)), 0
+                    FROM cp_pool pool
+                    WHERE pool.owner_id = :owner_id
+                    AND pool.startDate <= :start_date AND pool.endDate >= :end_date
+                    UNION
+                    SELECT CAST(pcmap.child_uuid AS CHAR(32)), pgraph.depth + 1
+                    FROM parent_child_map pcmap
+                    JOIN product_graph pgraph ON pgraph.uuid = pcmap.parent_uuid
+                    WHERE pgraph.depth < 10
+                )
+                SELECT content.uuid, pc.enabled
+                FROM product_graph pgraph
+                JOIN cp_product_contents pc ON pc.product_uuid = pgraph.uuid
+                JOIN cp_contents content ON content.uuid = pc.content_uuid
+            """;
+
+        List<Object[]> results = entityManager.createNativeQuery(sql)
             .setParameter("owner_id", ownerId)
             .setParameter("start_date", now)
             .setParameter("end_date", now)
-            .getResultList());
+            .getResultList();
 
-        Query prodQuery = entityManager.createNativeQuery(sql);
-        int blockSize = Math.min(this.getInBlockSize(), this.getQueryParameterLimit() / 2);
-
-        // Get all of the derived products and provided products for the products we've found, recursively
-        for (Set<String> lastBlock = productUuids; !lastBlock.isEmpty();) {
-            Set<String> children = new HashSet<>();
-
-            for (List<String> block : this.partition(lastBlock, blockSize)) {
-                children.addAll(prodQuery.setParameter("product_uuids", block)
-                    .getResultList());
-            }
+        List<ProductContent> activeContent = new ArrayList<>();
 
-            productUuids.addAll(children);
-            lastBlock = children;
-        }
-
-        // Use the product UUIDs to select all related enabled content
-        jpql = "SELECT DISTINCT pc.content, pc.enabled FROM ProductContent pc " +
-            "WHERE pc.product.uuid IN (:product_uuids)";
-
-        Query contentQuery = entityManager.createQuery(jpql);
-        HashMap<Content, Boolean> activeContent = new HashMap<>();
-
-        for (List<String> block : this.partition(productUuids)) {
-            contentQuery.setParameter("product_uuids", block)
-                .getResultList()
-                .forEach(col -> {
-                    Content content = (Content) ((Object[]) col)[0];
-                    Boolean enabled = (Boolean) ((Object[]) col)[1];
-                    activeContent.merge(content, enabled, (v1, v2) ->
-                        v1 != null && v1.booleanValue() ? v1 : v2);
-                });
+        for (Object[] result : results) {
+            Content content = this.getByUuid((String) result[0]);
```

Comment:
This will N+1 us to death. When I mentioned "another query" I meant doing the secondary lookup in bulk. How this will end up looking is something like this:

```
List<ProductContent> output = new ArrayList<>();
Map<String, Boolean> contentEnablementMap = entityManager.createNativeQuery(sql)
  .setParameter(...)
  .getResultStream()
  .collect(Collectors.toMap(row -> (String) row[0], row -> (boolean) row[1], (ele1, ele2) -> ele1 || ele2));

String contentLookupJPQL = "SELECT content FROM Content content " +
  "WHERE content.uuid IN (:content_uuids)";

TypedQuery<Content> contentQuery = entityManager.createQuery(contentLookupJPQL, Content.class);

for (List<String> block : this.partition(contentEnablementMap.keySet())) {
  contentQuery.setParameter("content_uuids", block)
    .getResultStream()
    .map(content -> new ProductContent(content, contentEnablementMap.get(content.getUuid())))
    .forEach(output::add);
}

return output;
```

Note that there are a few decisions to make in there (map construction, collision resolution, how the output should be collected, etc.); some of which should be properly documented so there aren't any surprises for callers.

Reply:
done